### ISSUE TITLE
Wrong import pattern of dolfin.io

### ISSUE THREAD

**Issue by michalhabera** [FEniCS/dolfinx#124]
The file `python/dolfin/io.py` appends just some (read_checkpoint, ) methods of `HDF5File` and `XDMFFile` classes.

If user doesn't do `import dolfin.io` then `python/dolfin/io.py` is not run and methods are not appended. This is not picked in unit tests since unit tests are run globally and during collection phase `test_HDF5_series.py` imports `dolfin.io` which makes everything look OK. 

Running purely `pytest test_XDMF.py -vsx` fails.

This can be solved e.g. adding `import io` to `dolfin/__init__.py`. Proper solution would be to wrap both XDMFFile and HDF5File classes in python layer with docs and import them.

**Comment by blechta** [comment]
Proper fix would be not to monkey-patch `dolfin.cpp` from `dolfin.io` but put these into `dolfin.cpp`.

**Comment by michalhabera** [comment]
@jan What do you mean? `dolfin.cpp` is an external library compiled to `.so` and generated exclusively by pybind11 wrappers.

If I understood it correctly the current approach is to wrap everything with own python-specific docs in `dolfin/` dir. See https://github.com/FEniCS/dolfinx/issues/20#issuecomment-386567025

**Comment by blechta** [comment]
Yes, the code in `python/dolfin/io.py` is clearly monkey-patching `dolfin.cpp`. I am saying don't do monkey-patching. Implement the functionality directly in `dolfin.cpp`. I think that pybind11 is flexible enough to allow that.

**Comment by michalhabera** [comment]
I understand, pybind is flexible enough. But then you need to do casting as here https://github.com/FEniCS/dolfinx/blob/master/python/src/io.cpp#L332 

In addition, documentation would need to go to pybind11 strings and that could be terrible. I agree with @garth 's original idea not to make `.cpp` public, but using inheritance monkey-patch and document pybind11 generated methods and classes. Monkey-patching is needed due to complicated inheritance of some classes. Again, we need to write a `Function` that inherits from `ufl.Coefficient`.

**Comment by blechta** [comment]
Then don't do `from .cpp.io import XDMFFile` in `dolfin/__init__.py`. Modify `XDMFFile` as you need in `dolfin.io` and do `from .io import XDMFFile`.

**Comment by michalhabera** [comment]
Yes ;) That I wrote as a proper solution in issue description, "Proper solution would be to wrap both XDMFFile and HDF5File classes in python layer with docs and import them.".

Its easy fix, I'll do it later, after my big PR to IO is done.

**Comment by garth-wells** [comment]
I don't see the issue; if one wants functionality from `dolfin.io`, then import  `dolfin.io`.

Load of the imports from `dolfin/__init__.py` will be removed. Many (all?) are very bad practice.

**Comment by jan** [comment]
@michalhabera I don't think you want to summon me from the depths.

**Comment by michalhabera** [comment]
Sorry @jan  . 
@garth-wells Currently, 
```python
from dolfin import XDMFFile
```
 imports XDMFFile without `.read_checkpoint()` method while 
```python
from dolfin import XDMFFile
import dolfin.io
```
works ok. This is a little issue of current state, not future design.
